src/operations/blas3/trsm.hpp

```diff
@@ -0,0 +1,158 @@
+SYCL_BLAS_INLINE bool
+DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::valid_thread(
+    cl::sycl::nd_item<1> id) const {
+  return true;
+}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+SYCL_BLAS_INLINE void DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::bind(
+    cl::sycl::handler& cgh) {
+  A_.bind(cgh);
+  invA_.bind(cgh);
+}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+SYCL_BLAS_INLINE void DiagonalBlocksInverter<
+    UnitDiag, Upper, matrix_t>::adjust_access_displacement() {
+  A_.adjust_access_displacement();
+  invA_.adjust_access_displacement();
+}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+template <typename local_memory_t>
+SYCL_BLAS_INLINE void DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::eval(
+    local_memory_t localMem, cl::sycl::nd_item<1> item) noexcept {
+  auto A = A_.get_data().get_pointer() + A_.get_access_displacement();
+  auto invA = invA_.get_data().get_pointer() + invA_.get_access_displacement();
+  value_t* local = localMem.localAcc.get_pointer();
+
+  const size_t threadIndex = item.get_local_id(0);
+  const size_t i = threadIndex;
+  const size_t blockIndex = item.get_group(0);
+
+  // Sets the offset for this particular block in the source and destination
+  // matrices
+  const size_t blockIndexPerBlock = blockIndex * internalBlockSize;
+  const size_t srcBlockOffset =
+      blockIndex * (internalBlockSize + lda_ * internalBlockSize);
+  const size_t numInnerBlocks = outterBlockSize / internalBlockSize;
+  const size_t blockIndexDiv = blockIndex / numInnerBlocks;
+  const size_t blockIndexMod = blockIndex % numInnerBlocks;
+  // go to the blockIndexDiv outer outterBlockSize*outterBlockSize block
+  const size_t offsetPart1 = blockIndexDiv * outterBlockSize * outterBlockSize;
+  // then to the blockIndexMod inner internalBlockSize*internalBlockSize block
+  // inside that
+  const size_t offsetPart2 =
+      blockIndexMod * (outterBlockSize * internalBlockSize + internalBlockSize);
+  const size_t destBlockOffset = offsetPart1 + offsetPart2;
+
+  // Loads the source lower triangle into local memory. Any values in the upper
+  // triangle or outside of the matrix are set to zero
+  for (size_t j = 0; j < internalBlockSize; ++j) {
+    bool isInRange = false;
+    if (upper) {
+      isInRange = (i <= j) && (blockIndexPerBlock + j < N_);
+    } else {
+      isInRange = (i >= j) && ((blockIndexPerBlock + i) < N_);
+    }
+    if (isInRange) {
+      const size_t srcIndex = j * lda_ + i + srcBlockOffset;
+      local[j + i * internalBlockSize] = A[srcIndex];
+    } else {
+      local[j + i * internalBlockSize] = value_t{0};
+    }
+  }
+  item.barrier(cl::sycl::access::fence_space::local_space);
+
+  // Inverts the diagonal elements
+  if (!unitDiag_) {
+    local[i + i * internalBlockSize] =
+        value_t{1} / local[i + i * internalBlockSize];
+    item.barrier(cl::sycl::access::fence_space::local_space);
```

Comment:
I know the `unitDiag_` is compile-time value and all threads will go through the if since the `unitDiag_` is the same for all threads, but we have seen in the past that some compiler had an issue putting barrier inside the if statement, even it is trivial and all taking the same path. Try to see if this could cause a problem on devices that we are supporting

Reply:
The tests are supposed to fail if there is a problem with this barrier so we can see if this causes a problem

Reply:
yes, What I am saying is that have you run them internally to see if it is not failing. Since travis only test on one simple default CPU